### ISSUE TITLE
Desktop CSD support

### ISSUE THREAD

**Issue by refi64** [flutter/flutter#31373]
(accidentally filed at and since moved from https://github.com/google/flutter-desktop-embedding/issues/340)

It would be great if desktop embedding supported CSDs. Something to the effect of:

```dart
import 'package:csd/csd.dart';

//... 
Csd.enableIfPossible(); // inside main()
// inside a build function
AppBar(
  actions: [
    if (Csd.enabled)
      MyCloseButton(),
  ]
)
```

I'd be happy to contribute to this, but it seems to be [pending some upstream GLFW stuff](https://github.com/glfw/glfw/pull/1420), unless Flutter does the window drag / resize callbacks manually...

**Comment by stuartmorgan** [comment]
Context: https://en.wikipedia.org/wiki/Client-Side_Decoration

> I'd be happy to contribute to this, but it seems to be pending some upstream GLFW stuff

I'm not sure supporting this in the GLFW shell is useful since that won't be the primary supported shell on any platform. It's something to consider for the real Windows (likely UWP) and Linux (likely GTK) shells, however.

**Comment by refi64** [comment]
At least for GTK that would actually be far easier, since GDK already has functions for this that also take care of the X11 / Wayland differences.

Thinking about this a bit more, there would likely be some widget `CsdBar` or similar that would wrap a child and automatically call the platform-specific drag functions.

**Comment by handicraftsman** [comment]
I believe SDL shell would be pretty nice for this - it abstracts most platform specific features like this.

**Comment by stuartmorgan** [comment]
Abstracting the platform at the level of implementing the embedding isn't a goal. Embeddings are platform-specific implementations of our own abstraction. Additional layers of abstractions make it harder to implement platform-specific features, or to support uses like add-to-app.

**Comment by patriziobruno** [comment]
This is now on a deprecated milestone. Should it be moved to a new milestone? With Canonical helping on the Linux front it should be easier to implement, at least on that platform. Ubuntu is Gnome based and Gnome heavily relies on CSD.

**Comment by stuartmorgan** [comment]
Priorities replaced non-time-specific milestone.

**Comment by P-B1101** [comment]
any update for this feature?

**Comment by bitsdojo** [comment]
I created a package for this. You can watch a tutorial on how to use it here:

https://www.youtube.com/watch?v=bee2AHQpGK4

**Comment by ikeofkc** [comment]
@bitsdojo 
Is this a platform specific solution. I'm guessing it won't allow CSDs in Linux, which is now standard in GNOME desktop / GTK which is the default desktop on most mainstream distros. 

I'd actually be interested in developing a custom GNOME/Adwaita centric widget set that could help development for mobile+desktop apps for new linux mobile variants like pureOS, Phos shell, etc.  I think it would be important to have proper CSDs implemented in flutter first however.

**Comment by bitsdojo** [comment]
> @bitsdojo
> Is this a platform specific solution?

Yes, it is a platform specific solution. Just released preliminary support for macOS as well. Looking at Linux after macOS is complete.

**Comment by zhuxiujia** [comment]
Is there any progress on this feature？

**Comment by adnan449** [comment]
+1 for frameless windows support on GTK/Linux. What are the blocking issues here?

**Comment by bitsdojo** [comment]
Just added GTK/Linux support to https://pub.dev/packages/bitsdojo_window

**Comment by Elixell** [comment]
Any progress with the official solution? Because bitsdojo is no longer updated and crashes with the recent flutter version.

**Comment by bitsdojo** [comment]
> Any progress with the official solution? Because bitsdojo is no longer updated and crashes with the recent flutter version.

Been away in the last months due to medical issues but I am now recovering and working on a new release.

**Comment by damywise** [comment]
https://pub.dev/packages/window_manager
also works

**Comment by scotch-hub** [comment]
Really hope this features gets worked on in the future. Things like custom headerbars, window shadows and window transparency are essential in making flutter for the desktop more extensive and customizable.

**Comment by jdressu** [comment]
came to give flutter a try.. cus everyone seems to be saying that it has this massive 'potencial' and flexibility for multiple platform projects..
but then i find out that there is functionality that should've been added right from the start that's still missing after what (Apr 21, 2019)? 3.5 years? 🤷‍♂️👎

**Comment by nathanfranke** [comment]
Time flies

**Comment by orange-pig** [comment]
almost 4 years

**Comment by hellolxc** [comment]
It's been five years now. 🙂

**Comment by stuartmorgan** [comment]
Adjusting team to Framework, since the core design work here would be the framework-level API, after which platform-specific engine bugs could be filed to support that API.
